File: go/find_all_numbers_disappeared_in_array.go
package leetcode

// inplace, O(n) time, O(1) space
func findDisappearedNumbers(nums []int) []int {
	for _, el := range nums {
		var idx int
		if el < 0 {
			idx = -el
		} else {
			idx = el
		}
		if nums[idx-1] > 0 {
			nums[idx-1] *= -1
		}
	}
	var res []int
	for i, el := range nums {
		if el > 0 {
			res = append(res, i+1)
		}
	}
	return res
}

// using an map, O(n) time, O(n) space
// func findDisappearedNumbers(nums []int) []int {
//     dic := make(map[int]bool)
//
//     for _, el := range nums {
//         dic[el] = true
//     }
//     var res []int
//     for i := 1; i < len(nums)+1; i++ {
//         _, ok := dic[i]
//         if ok != true {
//             res = append(res, i)
//         }
//     }
//     return res
// }

File: python/find-all-numbers-disappeared-in-array.py
# Inplace, O(n) time, O(1) space
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        for i, el in enumerate(nums):
            if nums[abs(el)-1] > 0:
                nums[abs(el)-1] *= -1
        
        return [i+1 for i in range(len(nums)) if nums[i] > 0]

# Using a set, O(n) time and O(n) space    
# class Solution:
#     def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
#         s = set(nums)
#         return [i for i in range(1,len(nums)+1) if i not in s]
                    